src/idl/src/annotation.c

```diff
@@ -187,47 +183,61 @@ annotate_extensibility(
   idl_annotation_appl_t *annotation_appl,
   idl_node_t *node)
 {
-  idl_extensibility_t extensibility = IDL_EXTENSIBILITY_FINAL;
+  const idl_annotation_appl_t **annotationp = NULL;
+  idl_extensibility_t *extensibilityp = NULL, extensibility = IDL_FINAL;
   const char *annotation;
 
   assert(annotation_appl);
   annotation = idl_identifier(annotation_appl->annotation);
   assert(annotation);
 
   if (strcmp(annotation, "final") == 0) {
-    extensibility = IDL_EXTENSIBILITY_FINAL;
+    extensibility = IDL_FINAL;
   } else if (strcmp(annotation, "appendable") == 0) {
-    extensibility = IDL_EXTENSIBILITY_APPENDABLE;
+    extensibility = IDL_APPENDABLE;
   } else if (strcmp(annotation, "mutable") == 0) {
-    extensibility = IDL_EXTENSIBILITY_MUTABLE;
+    extensibility = IDL_MUTABLE;
   } else if (strcmp(annotation, "extensibility") == 0) {
-    const char *enumerator = "";
+    const char *enumerator;
 
     assert(annotation_appl->parameters);
     assert(annotation_appl->parameters->const_expr);
     enumerator = idl_identifier(annotation_appl->parameters->const_expr);
     assert(enumerator);
     if (strcmp(enumerator, "FINAL") == 0) {
-      extensibility = IDL_EXTENSIBILITY_FINAL;
+      extensibility = IDL_FINAL;
     } else if (strcmp(enumerator, "APPENDABLE") == 0) {
-      extensibility = IDL_EXTENSIBILITY_APPENDABLE;
+      extensibility = IDL_APPENDABLE;
     } else if (strcmp(enumerator, "MUTABLE") == 0) {
-      extensibility = IDL_EXTENSIBILITY_MUTABLE;
+      extensibility = IDL_MUTABLE;
     }
   }
 
   if (idl_is_struct(node)) {
-    ((idl_struct_t *)node)->extensibility = extensibility;
+    annotationp = &((idl_struct_t *)node)->extensibility.annotation;
+    extensibilityp = &((idl_struct_t *)node)->extensibility.value;
   } else if (idl_is_union(node)) {
-    ((idl_union_t *)node)->extensibility = extensibility;
+    annotationp = &((idl_union_t *)node)->extensibility.annotation;
+    extensibilityp = &((idl_union_t *)node)->extensibility.value;
   } else if (idl_is_enum(node)) {
-    ((idl_enum_t *)node)->extensibility = extensibility;
-  } else {
+    annotationp = &((idl_enum_t *)node)->extensibility.annotation;
+    extensibilityp = &((idl_enum_t *)node)->extensibility.value;
+  }
+
+  assert((annotationp == NULL) == (extensibilityp == NULL));
+
+  if (!annotationp) {
     idl_error(pstate, idl_location(annotation_appl),
       "@%s can only be applied to constructed types", annotation);
     return IDL_RETCODE_SEMANTIC_ERROR;
+  } else if (*annotationp) {
+    idl_error(pstate, idl_location(annotation_appl),
+      "@%s clashes with an earlier annotation", annotation);
```

Comment:
Error messages indeed leave something to be desired. However, rather than fixing this single instance, I'd propose extending the respective functions to accept a custom flag that places a formatted location string in the message based on `idl_location_t` so it can be easily added in other places too. If agreed, creating a new issue for that might be the best way to proceed?

Reply:
Certainly